### PR TITLE
Simplified traits.

### PR DESCRIPTION
**Pull request by gabrik**
The idea behind this PR is to simplify the usage of Zenoh-Flow interface when creating nodes.

**Before:**
```rust



#[derive(Debug)]
struct SumAndSendFactory;

struct SumAndSend {
    input: Arc<Input>,
    output: Arc<Output>,
    counter: Arc<AtomicUsize>,
}

static INPUT: &str = "Number";
static OUTPUT: &str = "Sum";

#[async_trait]
impl OperatorFactoryTrait for SumAndSendFactory {
    async fn new_operator(
        &self,
        _context: &mut Context,
        _configuration: &Option<Configuration>,
        mut inputs: Inputs,
        mut outputs: Outputs,
    ) -> Result<Option<Arc<dyn Node>>> {
        let input = inputs.take_into_arc(INPUT).unwrap();
        let output = outputs.take_into_arc(OUTPUT).unwrap();
        let counter = Arc::new(AtomicUsize::new(0));

        Ok(Some(Arc::new(SumAndSend {
            input,
            output,
            counter,
        })))
    }
}

#[async_trait]
impl Node for SumAndSend {
    async fn iteration(&self) -> Result<()> {
        let data = if let Message::Data(mut message) = self.input.recv_async().await.unwrap() {
            message.get_inner_data().try_get::<ZFUsize>()?.clone()
        } else {
            bail!(ErrorKind::Empty, "No data")
        };

        self.counter.fetch_add(data.0, Ordering::AcqRel);
        let data = Data::from(ZFUsize(self.counter.load(Ordering::Acquire)));

        self.output.send_async(data, None).await
    }
}

zenoh_flow::export_operator_factory!(register);

fn register() -> Result<Arc<dyn OperatorFactoryTrait>> {
    Ok(Arc::new(SumAndSendFactory) as Arc<dyn OperatorFactoryTrait>)
}

```

**After**

```rust
#[export_operator]
struct SumAndSend {
    input: Arc<Input>,
    output: Arc<Output>,
    counter: Arc<AtomicUsize>,
}

static INPUT: &str = "Number";
static OUTPUT: &str = "Sum";

#[async_trait]
impl Operator for SumAndSend {

    fn new(
         _context: &mut Context,
        _configuration: &Option<Configuration>,
        mut inputs: Inputs,
        mut outputs: Outputs,
    )-> Result<Option<Self>> {
            Ok(Some(SumAndSend {
                input,
                output,
                counter,
            }))
    }
}

#[async_trait]
impl Node for SumAndSend {
    async fn iterartion(&self) -> Result<()> {
        let data = if let Message::Data(mut message) = self.input.recv_async().await.unwrap() {
            message.get_inner_data().try_get::<ZFUsize>()?.clone()
        } else {
            bail!(ErrorKind::Empty, "No data")
        };
        
        self.counter.fetch_add(data.0, Ordering::AcqRel);
        let data = Data::from(ZFUsize(self.counter.load(Ordering::Acquire)));

        self.output.send_async(data, None).await
    
    }
}

```

### CODE REVIEW

File: zenoh-flow/src/macros.rs [hunk 1]
```diff
@@ -65,40 +76,47 @@ macro_rules! export_operator_factory {
...
+                        Err(e) => Err(e),
+                    }
+                }),
             };
     };
 }
 
 /// This macros should be used in order to provide the symbols
-/// for the dynamic load of a Sink. Along with a register function
+/// for the dynamic load of an Source. Along with a register function
```
Comment (J-Loudet) [inline zenoh-flow/src/macros.rs]: s/an/a

File: zenoh-flow/src/traits.rs [hunk 2]
```diff
@@ -99,253 +99,310 @@ pub trait ZFData: DowncastAny + Debug + Send + Sync {
...
-/// Sources only possess `Outputs` and their purpose is to fetch data from the external world.
+/// A `CastSource` is an internal structure used to foster code reuse
+/// in the runtime. I.e., it allows all `Source`s to be `Node`s.
+pub(crate) struct CastSource(Arc<dyn Source>);
+
+impl From<Arc<dyn Source>> for CastSource {
+    fn from(other: Arc<dyn Source>) -> Self {
+        Self(other)
+    }
+}
```
Comment (J-Loudet) [inline zenoh-flow/src/traits.rs]: Having:
```rust
pub(crate) struct CastSource(pub(crate) Arc<dyn Source>);
```
would remove the need to have an `impl From` and allow us to do:
```rust
let cast_source = CastSource(source);
```

File: zenoh-flow/src/traits.rs [hunk 3]
```diff
@@ -99,253 +99,310 @@ pub trait ZFData: DowncastAny + Debug + Send + Sync {
...
-///
+/// A `CastSink` is an internal structure used to foster code reuse
+/// in the runtime. I.e., it allows all `Sink`s to be `Node`s.
+pub(crate) struct CastSink(Arc<dyn Sink>);
+
+impl From<Arc<dyn Sink>> for CastSink {
+    fn from(other: Arc<dyn Sink>) -> Self {
+        Self(other)
+    }
+}
```
Comment (J-Loudet) [inline zenoh-flow/src/traits.rs]: Same comment

File: zenoh-flow/src/traits.rs [hunk 4]
```diff
@@ -99,253 +99,310 @@ pub trait ZFData: DowncastAny + Debug + Send + Sync {
...
-/// }
+/// A `CastOperator` is an internal structure used to foster code reuse
+/// in the runtime. I.e., it allows all `Operator`s to be `Node`s.
+pub(crate) struct CastOperator(Arc<dyn Operator>);
+
+impl From<Arc<dyn Operator>> for CastOperator {
+    fn from(other: Arc<dyn Operator>) -> Self {
+        Self(other)
+    }
+}
```
Comment (J-Loudet) [inline zenoh-flow/src/traits.rs]: Same comment

File: zenoh-flow/src/traits.rs [hunk 5]
```diff
@@ -99,253 +99,310 @@ pub trait ZFData: DowncastAny + Debug + Send + Sync {
...
+        Sink::iteration(self).await
+    }
+}
+
+#[async_trait]
+impl Node for dyn Operator {
+    async fn iteration(&self) -> ZFResult<()> {
+        Operator::iteration(self).await
+    }
 }
```
Comment (J-Loudet) [inline zenoh-flow/src/traits.rs]: Being perfectly aware that this was my suggestion, I feel the need to ask: is this of any use in the end?

**Review by J-Loudet — COMMENTED** [review]
Small feedback

File: zenoh-flow/src/traits.rs (hunk 5 above)
Comment (gabrik) [reply zenoh-flow/src/traits.rs]: If commented it does not compile, so it is quite useful

**Review by J-Loudet — APPROVED** [review]
LGTM!

**Pull request merged**